accepted/0003-interactive-audit-resolver.md

```diff
@@ -0,0 +1,92 @@
+# Interactive audit resolver
+
+## Summary
+
+This proposal is for adding means for a human to resolve issues from `npm audit` and interactively make decisions about each issue. Available as `npm audit resolve` command.
+
+Features:
+- Remember the resolutions and allow them to affect `npm audit` behavior.
+- Decision options include 'ignore', 'remind later', 'fix', 'remove'.
+- Allow tracking who resolved an issue and when using git history.
+
+
+## Motivation
+
+At times, running `npm audit fix` won't fix all audit issues. Fixes might not be available or the user might not want to apply some of them, eg. major version updates or build/test dependencies.
+
+It should be possible to make `npm audit` a step in a CI pipeline to go red every time a new vulnerability affects the project. Managing security of dependencies should be quick to update, effective, easy to audit over time and secure in itself.
+
+For that, the user needs a new tool which makes addressing issues one by one convenient even across a whole ecosystem of projects.
+
+## Detailed Explanation
+
+`npm audit resolve` runs audit and iterates over all actions from the output with a prompt for a decision on each. 
+If fix is available, it's offered as one of the options to proceed. Other options include: ignore, remind later and delete.
+
+All decisions are stored in `audit-resolv.json` file as key-value, where key is `${dependency path}|${advisory id}` and value is a structure specific to the resolution.
+
+`npm audit` reads `audit-resolv.json` and respects the resolution (ignores ignored issues, ignores issues postponed to a date in the future)
+
+### resolutions
+- **fix** runs the fix proposed in audit action and marks the issue as fixed in `audit-resolv.json`. On each subsequent run of `npm audit resolve` if the issue comes up again, the user is also warned that the problem was supposed to be fixed already.
+- **ignore** marks the issue as ignored and `npm audit` will no longer fail because of it (but should display a single line warning about the issue being ignored). If a new vulnerability is found in the same dependency, it does not get ignored. If another dependency is installed with the same vulnerability advisory id it is not ignored. If the same package is installed as a dependency to another dependency (different path) it is not ignored.
+- **postpone** marks the issue as postponed to a timestamp 24h in the future. Instead of ignoring an issue permanently just to make a build pass, one can postpone it when in rush, but make it show up as a CI faiure on the next working day. 
```

Comment:
I thought about a few cases of audit trail with similar conclusions. IMHO the benefit of audit-resolv file's simplicity outweighs the inconvenience of using git history. Also, people are quite familiar with git so the less we do on our end, the less learning they have to do.

I'd like to avoid the file growing to overwhelming lengths, but other than that I'm fine with adding more capability to this file format if it's useful for the server side ignores etc.

Reply:
Leaving only the 24h option it should be annoying enough for most reasonable people to eventually fix. :)
Unless someone gets a job postponing vulnerabilities every morning before the team comes in.

With the audit-resolve format change I just pushed we can also add a count${dependency path}er. I'd consider that a nice touch, not the core functionality though.

Reply:
Update - instead of letting people postpone for more than 24h, due to a popular demand, I got an option in to make ignores expire. So if you decide to ignore something because there's no fix and it doesn't seem accessible from the code that depends on it, it's possible to make it expire.

Added `expiresAt` field to the decision object in resolve file